content/blog/2018/appswitch/index.md

```diff
@@ -0,0 +1,81 @@
+publishdate: 2018-11-28
+subtitle:
+attribution: Dinesh Subhraveti (AppOrbit and Columbia University)
+weight: 78
+keywords: [appswitch,performance]
+---
+
+We are going through an interesting cycle of application decomposition and recomposition.  While the microservice paradigm is driving monolithic applications to be broken into separate individual services, the service mesh approach is helping them to be connected back together into well-structured applications.  As such, microservices are logically separate but not independent.  They are usually closely interdependent and taking them apart introduces many new concerns such as need for mutual authentication between services.  Istio directly addresses most of those issues.
+
+## Dependency ordering problem
+
+An issue that arises due to application decomposition and one that Istio doesn’t address is dependency ordering -- bringing up individual services of an application in an order that guarantees that the application as a whole comes up quickly and correctly.  In a monolithic application, with all its components built-in, dependency ordering between the components is enforced by internal locking mechanisms.  But with individual services potentially scattered across the cluster in a service mesh, starting a service first requires checking that the services it depends on are up and available.
+
+Dependency ordering is deceptively nuanced with a host of interrelated problems.  Ordering individual services requires having the dependency graph of the services so that they can be brought up starting from leaf nodes back to the root nodes.  It is not easy to construct such a graph and keep it updated over time as interdependencies evolve with the behavior of the application.  Even if the dependency graph is somehow provided, enforcing the ordering itself is not easy.  Simply starting the services in the specified order obviously won’t do.  A service may have started but not be ready to accept connections yet.  This is the problem with docker-compose's `depends-on` tag, for example.
+
+Apart from introducing sufficiently long sleeps between service startups, a common pattern that is often used is to check for readiness of dependencies before starting a service.  In Kubernetes, this is typically done as part of the init container of the pod.  However that means that the entire application would be held up until all its dependencies come alive.  Sometimes applications spend several minutes initializing themselves on startup before making their first outbound connection.  Not allowing a service to start adds substantial overhead to overall startup time of the application.  Also, checking for readiness of dependencies won’t work in case of "multilevel" interdependencies where a service needs to be started before another service but only after a third service.  Likewise, multiple containers of the same pod may have mutual dependencies that cannot be ordered by simply waiting at the init container.
```

Comment:
> In Kubernetes, this is typically done as part of the init container of the pod.

This does not sound right.. Assuming you are talking about cross microservice dependencies, there is no way in Kubernetes to express this requirement. The readiness checks simply ensures that the instance of a service is up before it enters the load balancing pool. I feel that this paragraph is missing some connecting pieces or skipping ahead, as the flow changes from "service->service" initialization order to suddenly "intra-pod container start up order".. While theoretically you can stick all services of an application into a single pod, it leads to a different kind of problem than what you set out to talk about. Kubernetes has started addressing the container startup ordering issue within a single pod, but not the cross service initialization order.

Reply:
Actually I wasn't referring to k8s readiness probes...  ordering may be implemented with a script at the init container that waits on the dependencies.